tests/unit/lib/build_module/test_app_builder.py

```diff
@@ -41,21 +48,97 @@ def test_must_iterate_on_functions_and_layers(self):
         self.assertEqual(
             result,
             {
-                self.func1.name: build_function_mock.return_value,
-                self.func2.name: build_function_mock.return_value,
+                self.func1.name: os.path.join("builddir", self.func1.name),
+                self.func2.name: os.path.join("builddir", self.func2.name),
                 self.layer1.name: build_layer_mock.return_value,
                 self.layer2.name: build_layer_mock.return_value,
             },
         )
 
         build_function_mock.assert_has_calls(
             [
-                call(self.func1.name, self.func1.codeuri, self.func1.runtime, self.func1.handler, self.func1.metadata),
-                call(self.func2.name, self.func2.codeuri, self.func2.runtime, self.func2.handler, self.func2.metadata),
+                call(
+                    self.func1.name,
+                    self.func1.codeuri,
+                    self.func1.runtime,
+                    self.func1.handler,
+                    ANY,
+                    self.func1.metadata,
+                ),
+                call(
+                    self.func2.name,
+                    self.func2.codeuri,
+                    self.func2.runtime,
+                    self.func2.handler,
+                    ANY,
+                    self.func2.metadata,
+                ),
             ],
             any_order=False,
         )
 
+    @patch("samcli.lib.build.build_graph.BuildGraph._write")
+    def test_should_generate_build_graph(self, persist_mock):
+        build_graph = self.builder._get_build_graph()
+
+        self.assertTrue(len(build_graph.get_build_definitions()), 2)
+
+        all_functions_in_build_graph = []
+        for build_definition in build_graph.get_build_definitions():
+            for function in build_definition.functions:
+                all_functions_in_build_graph.append(function)
+
+        self.assertTrue(self.func1 in all_functions_in_build_graph)
+        self.assertTrue(self.func2 in all_functions_in_build_graph)
+
+    def test_should_run_build_for_only_unique_builds(self):
+        build_function_mock = Mock()
+
+        # create 3 function resources where 2 of them would have same codeuri, runtime and metadata
+        function1_1 = generate_function("function1_1")
+        function1_2 = generate_function("function1_2")
+        function2 = generate_function("function2", runtime="different_runtime")
+        resources_to_build_collector = ResourcesToBuildCollector()
+        resources_to_build_collector.add_functions([function1_1, function1_2, function2])
+
+        with osutils.mkdir_temp() as temp_base_dir:
```

Comment:
Shouldn't osutils.mkdir_temp be patched and mocked for unit tests? This also goes for other IO calls.

Reply:
Sure, I verified that reading and writing to `build.toml` file has already been tested in another file, I will mock here.